e2e/nomostest/config-sync.go

```diff
@@ -1255,96 +1257,147 @@ func resetRepository(nt *NT, repoType RepoType, nn types.NamespacedName, upstrea
 		nt.RootRepos[rsName] = resetRepository(nt, RootRepo, RootSyncNN(rsName), upstream, sourceFormat)
-		if err == nil {
-			nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceFormat": "%s", "git": {"dir": "%s"}}}`, sourceFormat, AcmeDir))
-		} else {
-			rs.Spec.SourceFormat = string(sourceFormat)
-			rs.Spec.Git = &v1beta1.Git{
-				Repo:      nt.GitProvider.SyncURL(nt.RootRepos[rsName].RemoteRepoName),
-				Branch:    MainBranch,
-				Dir:       AcmeDir,
-				Auth:      "ssh",
-				SecretRef: v1beta1.SecretReference{Name: controllers.GitCredentialVolume},
-			}
-			if err = nt.Create(rs); err != nil {
-				nt.T.Fatal(err)
-			}
+		defaultRS := RootSyncObjectV1Beta1FromRootRepo(nt, rsName)
+		if err = nt.Create(defaultRS); err != nil {
+			nt.T.Fatal(err)
 		}
+		return
+	}
+
+	defaultRS := RootSyncObjectV1Beta1FromRootRepo(nt, rsName)
+	if cmp.Equal(rs.Spec, defaultRS.Spec) {
+		// If found and unchanged, just reset the Repository.
+		// This can happen if the RootSync was not modified by the test.
+		nt.RootRepos[rsName] = resetRepository(nt, RootRepo, RootSyncNN(rsName), upstream, sourceFormat)
+		return
 	}
+
+	// If found and needs changes, we need to pause reconciliation to avoid a
+	// race condition between the reconciler-manager and the reconciler.
+
+	// Pause the reconciler-manager.
+	// This allows us to reset the RootSync Git config without updating the reconciler.
+	scaleDeployment(nt, reconcilermanager.ManagerName, configmanagement.ControllerNamespace, 0)
+	// Pause the reconciler.
+	// This allows us to reset the Git repo without syncing.
+	scaleDeployment(nt, core.RootReconcilerName(rsName), configmanagement.ControllerNamespace, 0)
+	nt.RootRepos[rsName] = resetRepository(nt, RootRepo, RootSyncNN(rsName), upstream, sourceFormat)
+
+	// Reset the RootSync spec to default
+	defaultRS.Spec.DeepCopyInto(&rs.Spec)
+	if err = nt.Update(rs); err != nil {
+		nt.T.Fatal(err)
+	}
+	// nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceFormat": "%s", "git": {"dir": "%s"}}}`, sourceFormat, AcmeDir))
+
+	// Resume the reconciler-manager, which will update and scale up the reconciler.
+	scaleDeployment(nt, reconcilermanager.ManagerName, configmanagement.ControllerNamespace, 1)
 }
 
-// resetRootRepos resets all RootSync objects to the initial commit to delete
-// managed testing resources.
-//
-// The spec.sourceFormat is reset to the specified format, to ensure the
-// remaining objects are deleted without erroring from format mismatch.
-//
-// The spec.git.dir is reset to "acme", which is created by
-// NewRepository and Repository.ReInit and must still exist.
-//
-// TODO: reset everything else, including the branch
-func resetRootRepos(nt *NT, upstream string, sourceFormat filesystem.SourceFormat) {
-	// Reset the default root repo first so that other managed RootSyncs can be re-created and initialized.
+func resetNamespaceRepo(nt *NT, rsNN types.NamespacedName, upstream string, sourceFormat filesystem.SourceFormat) {
```

Comment:
Would be a bit cooler if this was generic instead of the a duplicate of resetRootRepo

Reply:
Same problem everywhere, no abstraction around RootSync/RepoSync...

Also, in this case the RootRepos are keyed by name and the NonRootRepos are keyed by NamespacedName. That needs a huge refactor to change.

Reply:
😢